extra/camino/man/man1/estimatesnr.1
.\" $Id$

.TH estimatesnr 1

.SH NAME
estimatesnr \- Estimates the signal to noise ratio in b=0 images.

.SH SYNOPSIS
\fBestimatesnr\fR \fB-bgmask\fR <\fImask\fR> \fB-schemefile\fR <\fIfile\fR>
[\fB-noiseroi\fR <\fIroi\fR>]

.SH DESCRIPTION

This program estimates the noise variance and the signal to noise ratio in a given ROI,
for the b=0 data for a given scan.

The traditional method for estimating the noise is to sample two ROIs, one in brain white
matter, and one in the background. Assuming that the background signal contained only
noise, we can estimate the noise standard deviation as

  sigma = sqrt(2.0 / (4.0 - PI)) * stddev(signal in background region)

where the constant scaling corrects for the Rician distribution of the noise, giving us
the standard deviation sigma of the original signal. To synthesize data with the same
noise conditions, we would take the true signal S_0 and calculate

  S = |[S_0 + N(0, sigma), N(0, sigma)]|

where N(0, sigma) is a random sample drawn from normal distribution with mean 0 and
standard deviation sigma.

The two-ROI approach is only valid when the signal in the background follows the
distribution

  S_b = |[N(0, sigma), N(0, sigma)]|

however, this is often not the case. Dietrich et al [JMRI 26:376-385 (2007)] showed the
shortcomings of this approach and evaluated better ways to estimate the noise variance
without using a background region.

.SH ALTERNATIVES TO THE TWO-ROI APPROACH

The first method requires two b=0 images. We call this method sigma_diff, because it uses
the difference between the two images. If S_{i1} is the signal in voxel i, image 1, and
S_{i2} is the signal from the same voxel in image 2, and there are N voxels in the ROI,
then

  sigma_diff = stddev(S_{11} - S_{12},...,S_{N1} - S_{N2}) / sqrt(2)

The SNR is the mean signal in the ROI divided by sigma_diff, ie

  mean(S_{i1} + S_{i2}) / [2.0 * sigma_diff]

The second method requires multiple b=0 images, and defines sigma_mult as the standard
deviation of the signal over the ROI, across all K b=0 images. Again, let i be a voxel
index, then

  sigma_i = stddev(S_{i1},...,S{iK}))

  sigma_mult = mean(sigma_1,...,sigma_N)

And finally SNR is

  mean(S_{11}, S_{12},...,S_{1K}, S_{21},...,S_{NK}) / sigma_mult

If there are two or more b=0 images, both snr_diff and snr_mult will both be computed.
The more b=0 images there are, the better the estimate via sigma_mult, but sigma_diff
only ever uses the first two b=0 images.

.SH USES OF NOISE VARIANCE IN CAMINO

The noise variance, or the signal to noise ratio (SNR) in white matter, is used in
several Camino programs, such as:

\fBpicopdfs\fR - the SNR is required to evaluate the uncertainty in principal directions.

\fBrestore\fR - the noise standard deviation is required to do the fitting. However, the
RESTORE method contains other parameters which have the same effect as altering the
estimate of sigma. Thus you should consider sigma as a starting point for the RESTORE
algorithm.

\fBmbalign\fR - this program uses RESTORE, see above.

.SH EXAMPLES

Data exists in a voxel-ordered file, A.Bfloat. A white-matter region WM_ROI.hdr is drawn
by hand.

  estimatesnr -inputfile A.Bfloat -schemefile A.scheme -bgmask WM_ROI.hdr

Now we compute the same thing, only we use BG_ROI.hdr, which is simply WM_ROI.hdr
translated such that the non-zero voxels are all in the background region of the image.

  estimatesnr -inputfile A.Bfloat -schemefile A.scheme -bgmask WM_ROI.hdr -noiseroi BG_ROI.hdr

Start from data in a 4D Analyze file

  analyze2voxel -inputfile data.hdr | estimatesnr -inputfile A.Bfloat -schemefile A.scheme -bgmask WM_ROI.hdr

.SH OPTIONS
.TP
.B \-bgmask\fR <\fImask\fR> 
Path to an image header or raw data file that contains the region of interest. This image
should mask  out a region of interest that contains an approximately constant unweighted
signal. Usually, we would choose a white matter structure for this purpose.

.TP
.B \-schemefile\fR <\fIfile\fR> 
Imaging scheme for the data. The program will select the b=0 measurements for the
computation.

.TP
.B \-noiseroi\fR <\fIroi\fR> 
An ROI placed outside the brain, where the mean signal is zero. If this is supplied, the
noise variance is estimated using the "two ROI" technique.

.TP
.B \-approxb0meas\fR <\fIx y z...\fR> 
Tells the program to use the specified measurements instead of the b=0 measurements. The
arguments following The option should be integer indices of measurements that are
approximately directionally isotropic. For example, imagine data with no b=0
measurements, but the first three are at b=50 s / mm^2. The program could then be called
with the option "-approxb0meas 0 1 2". This option should only be used when signal is
homogeneous across the selected measurements and across the ROI. This will not be the
case for most diffusion-weighted measurements in the brain.

.SH "AUTHORS"
Philip Cook <camino@cs.ucl.ac.uk>

.SH SEE ALSO
picopdfs(1), restore(1), mbalign(1), datasynth(1), addnoise(1), analyze2voxel(1)

.SH BUGS


extra/camino/apps/EstimateSNR.java
package apps;

import data.*;
import imaging.*;
import misc.*;
import tools.*;

import java.io.*;
import java.text.DecimalFormat;
import java.util.ArrayList;



/**
 *
 * Estimates SNR using different methods explained in the paper "Measurement of signal-to-noise ratios in MR
 * images: Influence of multichannel coils, parallel imaging, and reconstruction filters" by Olaf Dietrich
 * et al, JMRI 26:376-385 (2007).
 * <p>
 * The standard method is to divide the mean of a tissue ROI by the standard deviation of a background ROI, but
 * this can be unreliable with modern imaging techniques. We therefore examine the signal variance across multiple
 * b=0 images to estimate the SNR.
 * </p>
 *
 * @author Philip Cook
 * @version $Id$
 *
 */
public class EstimateSNR extends Executable{
    
    //public static void main(String[] args) {
    public EstimateSNR(String[] args){
        super(args);
    }
    
    DataSource bgROI_Source;
    int[] b0Indices;
    DW_Scheme scheme;
	int numMeas;
	
    public void initDefaultVals(){
        bgROI_Source = null;
        b0Indices = null;
        
    }
    
    public void initOptions(String[] args){
              
        CL_Initializer.inputDataType = "float";
        CL_Initializer.CL_init(args);
        
        CL_Initializer.initImagingScheme();
        CL_Initializer.initDataSynthesizer();
		
		for (int i = 0; i < args.length; i++) {
            
            if(args[i].equals("-noiseroi")) {
                try {
                    bgROI_Source = ImageHeader.readHeader(args[i+1]).getImageDataSource();
                }
                catch (IOException e) {
                    throw new LoggedException("Cannot read background ROI " + args[i+1]);
                }
                
                CL_Initializer.markAsParsed(i,2);
            }
            if (args[i].equals("-approxb0meas")) {
                String[] tmp = new String[1000];
                int t = 0;
                
                while (t + i + 1 < args.length && !args[i+t+1].startsWith("-")) {
                    tmp[t] = args[i+t+1];
                    t++;
                }
                
                b0Indices = new int[t];
                
                for (int index = 0; index < t; index++) {
                    b0Indices[index] = Integer.parseInt(tmp[index]);
                }
                
                CL_Initializer.markAsParsed(i, t+1);
            }
            
        }
        
        CL_Initializer.checkParsing(args);
        
        if (CL_Initializer.bgMask == null) {
            throw new LoggedException("This program requires a foreground ROI with -bgmask");
        }
        
    }
        
	public void initVariables(){	
        scheme = CL_Initializer.imPars;
		numMeas = scheme.numMeasurements();
    }   
	
    public void execute(OutputManager om){
        // flat list of foreground data in voxel order, ie all measurements from the
        // same voxel are together
        ArrayList<Double> fg = new ArrayList<Double>();
        
        ArrayList<Double> bg = new ArrayList<Double>();
        
//        int numMeas = scheme.numMeasurements();
        
        boolean[] isB0 = new boolean[numMeas];
        
        int numB0 = 0;
        
        if (b0Indices != null) {
            
            // override scheme file and treat specified indices as b=0
            
            numB0 = b0Indices.length;
            
            for (int i = 0; i < numB0; i++) {
                isB0[b0Indices[i]] = true;
            }
        }
        else {
            for (int i = 0; i < numMeas; i++) {
                isB0[i] = scheme.zero(i);
                
                if (isB0[i]) {
                    numB0++;
                }
            }
            
        }
        if (numB0 == 0) {
            throw new LoggedException("Can't calculate SNR without any data at b=0");
        }
        
        
        // number of voxels in the background ROI
        int bgVoxels = 0;
        
        int fgVoxels = 0;
        
        while (CL_Initializer.data.more()) {
            double[] voxel = CL_Initializer.data.nextVoxel();
            
            // check if voxel is in foreground ROI
            if (CL_Initializer.bgMask.nextVoxel()[0] > 0.0) {
                
                double[] b0 = new double[numB0];
                
                fgVoxels++;
                
                for (int i = 0; i < numMeas; i++) {
                    if (isB0[i]) {
                        fg.add(new Double(voxel[i]));
                    }
                }
                
            }
            
            // check if voxel is in the background ROI
            if (bgROI_Source != null && bgROI_Source.nextVoxel()[0] > 0.0) {
                // use first measurement for bg
                bg.add(new Double(voxel[0]));
                bgVoxels++;
            }
            
        }
        
        
        // now got data, convert to arrays
        double[][] data = new double[fgVoxels][numB0];
        
        Double[] tmp = new Double[fgVoxels*numB0];
        
        fg.toArray(tmp);
        
        int nextFG = 0;
        
        for (int i = 0; i < fgVoxels; i++) {
            for (int j = 0; j < numB0; j++) {
                data[i][j] = tmp[nextFG++].doubleValue();
            }
        }
        
        double[] bgData = null;
        
        if (bgROI_Source != null) {
            
            tmp = new Double[bgVoxels];
            
            bg.toArray(tmp);
            
            bgData = new double[bgVoxels];
            
            for (int i = 0; i < bgVoxels; i++) {
                bgData[i] = tmp[i].doubleValue();
            }
        }
        
        
        System.out.println("\nNumber of b=0 images:\t" + numB0 + "\n");
        
        System.out.println("\nNumber of foreground voxels:\t" + fgVoxels);
        System.out.println("\nNumber of background voxels:\t" + bgVoxels + "\n");
        
        System.out.println("\nSignal to noise in foreground ROI\n");
        
        double[] snrStdv = null;
        double[] snrDiff = null;
        double[] snrMult = null;
        
        if (bgData != null) {
            
            snrStdv = snrStdv(data, bgData);
        }
        if (numB0 > 1) {
            snrDiff = snrDiff(data);
            snrMult = snrMult(data);
        }
        
        DecimalFormat df = new DecimalFormat("0.00");
        
        if (snrStdv != null) {
            System.out.println("SNR stdv:\t\t" + df.format(snrStdv[0] / snrStdv[1]));
        }
        
        if (numB0 == 1) {
            System.out.println("SNR mult:\t\t undefined");
            System.out.println("SNR diff:\t\t undefined");
        }
        else {
            
            System.out.println("SNR diff:\t\t" + df.format(snrDiff[0] / snrDiff[1]));
            System.out.println("SNR mult:\t\t" + df.format(snrMult[0] / snrMult[1]));
        }
        
        System.out.println("\nNoise standard deviation\n");
        
        if (snrStdv != null) {
            System.out.println("sigma stdv:\t\t" + df.format(snrStdv[1]));
        }
        
        if (numB0 == 1) {
            System.out.println("sigma mult:\t\t undefined");
            System.out.println("sigma diff:\t\t undefined");
        }
        else {
            System.out.println("sigma diff:\t\t" + df.format(snrDiff[1]));
            System.out.println("sigma mult:\t\t" + df.format(snrMult[1]));
        }
        
        System.out.println();
    }
    
    
    
    /**
     * Computes SNR from a single ROI, imaged twice.
     *
     * @param data the data, dimensions [voxels][2].
     *
     * @return {S_{diff}, \sigma_{diff}}.
     */
    public static double[] snrDiff(double[][] data) {
        
        int N = data.length;
        
        // sMean = (1 / 2N) (data[i][0] + data[i][1]), 0 <= i < N, which is S_diff in the Dietrich paper
        double sMean = 0.0;
        
        // sDiff[i] = (data[i][0] - data[i][1])
        double[] sDiff = new double[N];
        
        double sigmaDiff = 0.0;
        
        for (int i = 0; i < N; i++) {
            // actually 2 * N * sMean here
            sMean += data[i][0] + data[i][1];
            
            sDiff[i] = data[i][0] - data[i][1];
        }
        
        sMean = sMean / (2.0 * N);
        
        sigmaDiff = ( 1.0 / Math.sqrt(2.0) ) * Math.sqrt(ArrayOps.var(sDiff, ArrayOps.mean(sDiff)));
        
        return new double[] {sMean, sigmaDiff};
    }
    
    
    /**
     * Computes SNR from a single ROI, imaged many times.
     *
     * @param data the data, dimensions [voxels][K] where there are K measurements.
     *
     * @return {S_{mult}, \sigma_{mult}}.
     */
    public static double[] snrMult(double[][] data) {
        
        // number of voxels
        int N = data.length;
        
        // number of measurements
        int K = data[0].length;
        
        // mean signal over all measurements and all voxels
        double sMult = 0.0;
        
        double[] std = new double[N];
        
        for (int i = 0; i < N; i++) {
            
            // normalize later
            sMult += ArrayOps.sum(data[i]);
            
            std[i] = Math.sqrt(ArrayOps.var(data[i], ArrayOps.mean(data[i])));
        }
        
        sMult = sMult / (N * K);
        
        return new double[] {sMult, ArrayOps.mean(std)};
        
    }
    
    
    /**
     * Computes SNR from a single ROI with noise variance estimated from background.
     *
     * @param fg the data for the foreground ROI, with dimension [voxels][1].
     * @param bg the data for the background ROI.
     *
     * @return {S, \sigma}.
     */
    public static double[] snrStdv(double[][] fg, double[] bg) {
        
        double[] fg1D = new double[fg.length];
        
        for (int i = 0; i < fg.length; i++) {
            fg1D[i] = fg[i][0];
        }
        
        double fgMean = ArrayOps.mean(fg1D);
        
        double bgStd = Math.sqrt(2.0 / (4.0 - Math.PI)) * Math.sqrt(ArrayOps.var(bg, ArrayOps.mean(bg)));
        
        return new double[] {fgMean, bgStd};
    }
    
    
    
    
}